File: clients/app/lib/api/realtime.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

import 'client.dart';
import 'models.dart';

enum RealtimeStatus { connecting, open, closed }

/// Auto-reconnecting WebSocket to /api/v1/realtime (messages, typing, applications,
/// stories and wallet updates). Refreshes the access token when the server closes with 4401.
class RealtimeConnection {
  RealtimeConnection(this.api) {
    _connect();
  }

  final OvlApi api;
  final ValueNotifier<RealtimeStatus> status = ValueNotifier(RealtimeStatus.connecting);
  final StreamController<RealtimeEvent> _events = StreamController.broadcast();
  WebSocketChannel? _channel;
  StreamSubscription<dynamic>? _sub;
  Timer? _retry;
  int _attempts = 0;
  bool _stopped = false;

  Stream<RealtimeEvent> get events => _events.stream;

  Future<void> _connect() async {
    if (_stopped) return;
    final tokens = api.tokens.tokens;
    if (tokens == null) {
      status.value = RealtimeStatus.closed;
      return;
    }
    status.value = RealtimeStatus.connecting;
    final http = api.uri('/realtime', {'token': tokens.accessToken});
    final url = http.replace(scheme: http.scheme == 'https' ? 'wss' : 'ws');
    try {
      final channel = WebSocketChannel.connect(url);
      _channel = channel;
      await channel.ready;
      if (_stopped) {
        await channel.sink.close();
        return;
      }
      _attempts = 0;
      status.value = RealtimeStatus.open;
      _sub = channel.stream.listen(
        (data) {
          try {
            _events.add(RealtimeEvent.fromJson(jsonDecode(data as String) as Json));
          } catch (_) {
            // Unknown event shapes are ignored.
          }
        },
        onDone: () => _closed(channel.closeCode),
        onError: (_) => _closed(null),
        cancelOnError: true,
      );
    } catch (_) {
      _closed(null);
    }
  }

  Future<void> _closed(int? code) async {
    _sub?.cancel();
    _sub = null;
    _channel = null;
    status.value = RealtimeStatus.closed;
    if (_stopped) return;
    if (code == 4401 && !await api.refresh()) return;
    final delay = Duration(milliseconds: min(30000, 1000 * pow(2, _attempts++).toInt()));
    _retry = Timer(delay, _connect);
  }

  /// Reconnect now (e.g. when the app returns to the foreground).
  void reconnect() {
    if (_stopped || status.value != RealtimeStatus.closed) return;
    _retry?.cancel();
    _attempts = 0;
    _connect();
  }

  void sendTyping(String chatId) {
    _channel?.sink.add(jsonEncode({'type': 'typing', 'chatId': chatId}));
  }

  void close() {
    _stopped = true;
    _retry?.cancel();
    _sub?.cancel();
    _channel?.sink.close();
    status.value = RealtimeStatus.closed;
    _events.close();
  }
}

File: packages/sdk/src/realtime.ts
import type { RealtimeClientMessage, RealtimeEvent } from '@ovl/shared';
import type { OvlClient } from './client';

export type RealtimeStatus = 'connecting' | 'open' | 'closed';

/**
 * Auto-reconnecting WebSocket connection to /api/v1/realtime.
 * Works with the WebSocket global of browsers, React Native, Node 22+, Tauri and Capacitor.
 */
export class RealtimeConnection {
  private socket: WebSocket | null = null;
  private readonly listeners = new Set<(event: RealtimeEvent) => void>();
  private readonly statusListeners = new Set<(status: RealtimeStatus) => void>();
  private stopped = false;
  private attempts = 0;
  private timer: ReturnType<typeof setTimeout> | null = null;
  status: RealtimeStatus = 'connecting';

  constructor(private readonly client: OvlClient) {
    this.connect();
  }

  private url(token: string): string {
    const loc = (globalThis as { location?: { protocol: string; host: string } }).location;
    const origin = this.client.baseUrl || (loc ? `${loc.protocol}//${loc.host}` : '');
    return `${origin.replace(/^http/, 'ws')}/api/v1/realtime?token=${encodeURIComponent(token)}`;
  }

  private setStatus(status: RealtimeStatus) {
    this.status = status;
    for (const l of this.statusListeners) l(status);
  }

  private connect() {
    if (this.stopped) return;
    const tokens = this.client.tokens.get();
    if (!tokens) return this.setStatus('closed');
    this.setStatus('connecting');
    const socket = new WebSocket(this.url(tokens.accessToken));
    this.socket = socket;
    socket.onopen = () => {
      this.attempts = 0;
      this.setStatus('open');
    };
    socket.onmessage = (message) => {
      let event: RealtimeEvent;
      try {
        event = JSON.parse(String(message.data)) as RealtimeEvent;
      } catch {
        return;
      }
      for (const l of this.listeners) l(event);
    };
    socket.onclose = async (event) => {
      this.socket = null;
      this.setStatus('closed');
      if (this.stopped) return;
      // 4401: the access token expired — refresh it before reconnecting.
      if (event.code === 4401 && !(await this.client.refresh())) return;
      const delay = Math.min(30_000, 1000 * 2 ** this.attempts++);
      this.timer = setTimeout(() => this.connect(), delay);
    };
  }

  on(listener: (event: RealtimeEvent) => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  onStatus(listener: (status: RealtimeStatus) => void): () => void {
    this.statusListeners.add(listener);
    return () => this.statusListeners.delete(listener);
  }

  send(message: RealtimeClientMessage): void {
    if (this.socket?.readyState === WebSocket.OPEN) this.socket.send(JSON.stringify(message));
  }

  typing(chatId: string): void {
    this.send({ type: 'typing', chatId });
  }

  close(): void {
    this.stopped = true;
    if (this.timer) clearTimeout(this.timer);
    this.socket?.close(1000, 'Client closed');
    this.listeners.clear();
    this.statusListeners.clear();
  }
}
